File: align/align.clj
#!/usr/bin/env bb
;; tryna see if I can do better than align.pl as far as clarity and tersity go
;; attempt 1: clojure
(ns fuck
  ;; (:use [clojure.repl])
  (:require [clojure.string :as str]
            [clojure.java.io :as io]))

(defn id [datum] datum)

(def lines
  (if (empty? *command-line-args*)
    (doall (line-seq (java.io.BufferedReader. *in*)))
    (reduce concat (for [filename *command-line-args*]
                     (with-open [r (io/reader filename)]
                       (doall (line-seq r)))))))

;; for testing (repl fuckery, where you can't rely on argv) use this instead
;; (def lines (with-open [f (io/reader "file")] (slurp-lines f)))

(defn pad [s target-length datum]
  (concat s (repeat (max 0 (- target-length (count s))) datum)))

(defn pad-align-lines [lines]
  (let* [split-lines (mapv #(clojure.string/split % #"[,\s]+") lines)
         most-fields-in-line (reduce max (map count split-lines))
         max-field-widths (reduce #(mapv max %1 %2)
                                   (mapv #(pad % most-fields-in-line 0)
                                          (mapv #(mapv count %) split-lines)))
         padded-fields (fn [line] (map #(str/join "" (pad %1 %2 " "))
                                       line max-field-widths))]
    (map #(str/join " " (padded-fields %))  split-lines)))

(doall (for [x (pad-align-lines lines)] (println x)))

File: align/align.py
#!/usr/bin/env python
# tryna see if I can do better than align.pl as far as clarity and tersity go
# attempt 2: python
import sys, re

lines=[]
if len(sys.argv) != 1:
    for fn in sys.argv[1:]:
        with open(fn) as f:
            for line in f:
                lines.append(line.strip())
else:
    lines = map(str.strip, sys.stdin.readlines())

lines = [re.split(r'[,\s]+', line) for line in lines]
line_field_lengths=[[len(elt) for elt in line] for line in lines]
max_line_fields=max(map(len, lines))
column_max_lengths=[]
for i in line_field_lengths:
    ml = max(len(column_max_lengths), len(i))
    column_max_lengths += [0 for _ in range(ml - len(column_max_lengths))]
    i += [0 for _ in range(ml - len(i))]
    column_max_lengths = [max(x,y) for (x,y) in zip(column_max_lengths, i)]

for l in lines:
    padded = [a + (' ' * (b-len(a))) for (a, b) in zip (l, column_max_lengths)]
    print(' '.join(padded))
